README.md
# DADI Cache

[![npm (scoped)](https://img.shields.io/npm/v/@dadi/cache.svg?maxAge=10800&style=flat-square)](https://www.npmjs.com/package/@dadi/cache)
[![coverage](https://img.shields.io/badge/coverage-73%25-yellow.svg?style=flat)](https://github.com/dadi/cache)
[![Build Status](https://travis-ci.org/dadi/cache.svg?branch=master)](https://travis-ci.org/dadi/cache)
[![JavaScript Style Guide](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square)](http://standardjs.com/)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg?style=flat-square)](https://github.com/semantic-release/semantic-release)

> A caching layer supporting Redis and filesystem caching, used in core DADI products

* [Overview](#overview)
* [Install](#install)
* [Usage](#usage)
  * [Create Cache instance](#create-cache-instance)
  * [Add an item to the cache](#add-an-item-to-the-cache)
  * [Get an item from the cache](#get-an-item-from-the-cache)
  * [Example real world usage](#example-real-world-usage)
* [Configuration](#configuration)
  * [General Options](#general-options)
  * [Default Options](#default-options)
  * [Filesystem Caching](#filesystem-caching)
  * [Redis Caching](#redis-caching)
  * [Redis Cluster](#redis-cluster)
* [Cache Fallback](#cache-fallback)

## Overview

Removing the complexity involved in setting up two separate cache handlers for every project, DADI Cache can use either the filesystem or a Redis instance to store and retrieve content.

## Install

```shell
npm install @dadi/cache --save
```

## Usage

### Create Cache instance

```js
// require the module
const Cache = require('@dadi/cache')

// setup the options for caching
// defaults if nothing specified: 
// {
//   directory: {
//     enabled: true,
//     path: './cache'
//   },
//   redis: {
//     enabled: false
//   }
// }
const options = {
  "ttl": 3600,
  "directory": {
    "enabled": false,
    "path": "./cache/"
  },
  "redis": {
    "enabled": true,
    "host": "127.0.0.1",
    "port": 6379
  }
}

const cache = new Cache(options)
```

### Add an item to the cache

#### `set(key, data)`

Returns a Promise that returns an empty String if successful, otherwise an Error.

The `data` argument can be a String, Buffer or Stream.

```js
let key = 'test-cached-item'
let data = 'test data'

cache.set(key, data).then(() => {
  // do something
}).catch(err => {
  // Error
})
```

### Get an item from the cache

#### `get(key)`

Returns a Promise that returns a Stream of the cached data if the key exists or an Error if it does not exist.
The error message returned is "The specified key does not exist".

```js
let key = 'test-cached-item'

cache.get(key).then(stream => {
  // do something with the stream
}).catch(err => {
  // err === "The specified key does not exist"
})
```

### Example real world usage

```js
const express = require('express')
const app = express()
const dadiCache = require('@dadi/cache')
const cache = new dadiCache()

app.get(function (req, res) {
  let key = req.url

  cache.get(key).then(stream => {
    // cached data found for req.url
    res.setHeader('X-Cache', 'HIT')
    stream.pipe(res)
  }).catch(err => {
    // cached data not found for req.url
    let content = fetchContent()

    // cache the content
    cache.set(key, content).then(() => {
      res.setHeader('X-Cache', 'MISS')
      res.end(content)
    })
  })
})
```

## Configuraton

### General options

Property|Description|Default|Example
--------|-----------|-------|-------
ttl|The time, in seconds, after which cached data is considered stale|true|3600

### Default options

A DADI Cache instance can be created with no options, in which case the following options will be used:

```js
{
  "directory": {
    "enabled": true,
    "path": "./cache"
  },
  "redis": {
    "enabled": false
  }
}
```

### Filesystem caching

Property|Description|Default|Example
--------|-----------|-------|-------
enabled|If true, caching is enabled using the following settings|true|
path|The absolute or relative path to the directory for cache files|"./cache"|"/tmp/dadi-cache/"
extension| (optional) The extension to use for cache files| none | "json"
directoryChunkSize| (optional) If set, cache files are stored in a series of subdirectories based on the cache key| 0 | 5
autoFlush | If true, DADI Cache will clear cache files that are older than the specified TTL setting, at the interval specified by `autoFlushInterval` | false | true
autoFlushInterval | The period of time between clearing cache files (in seconds) | 300 | 1800

### Redis caching

A set of options for both file and Redis caching _must_ be provided if you intend to use Redis as the cache store. This allows DADI Cache
to [fallback to file caching](#cache-fallback) in the event of a Redis connection failure.

Property|Description|Default|Example
--------|-----------|-------|-------
enabled|If true, caching is enabled using the following settings|false|true
host|The hostname or IP address of the Redis server |"127.0.0.1"|"drum.redistogo.com"
port|The port of the Redis server|6379 |9092

```json
{
  "directory": {
    "enabled": true,
    "path": "./cache"
  },
  "redis": {
    "enabled": true,
    "host": "127.0.0.1",
    "port": 6379
  }
}
```

### Redis Cluster

Property|Description|Default|Example
--------|-----------|-------|-------
enabled|If true, caching is enabled using the following settings|false|true
cluster|If true, DADI Cache will connect caching is enabled using the following settings|false|true
scaleReads|Specify where to send queries, to the masters, slaves, or a combination. See [Read-Write Splitting](#read-write-splitting) |"master"|"all"
hosts|When `cluster: true`, Cache uses this array of hosts to connect. Each array item must contain a `host` and `port`.|| ```[{"host":"127.0.0.1", "port": 6379}, {"host":"127.0.0.1", "port": 6380}]```

To connect to a Redis cluster an array of hosts must be specified, rather than a single host and port.

> The array does not need to contain all your cluster nodes, but a few so that if one is unreachable the next one will be tried. DADI Cache will discover other nodes automatically when at least one node is connnected.

```json
{
  "directory": {
    "enabled": true,
    "path": "./cache"
  },
  "redis": {
    "enabled": true,
    "cluster": true,
    "scaleReads": "all",
    "hosts": [
      {
        "host": "127.0.0.1",
        "port": 6379
      },
      {
        "host": "127.0.0.1",
        "port": 6383
      }
    ]
  }
}
```

#### Read-Write Splitting

A typical Redis cluster contains three or more masters and several slaves for each master. It's possible to scale out Redis cluster by sending read queries to slaves and write queries to masters by setting the `scaleReads` option.

`scaleReads` is "master" by default, which means no queries will be sent to slaves. The other available options:

* "all": Send write queries to masters and read queries to masters or slaves randomly.
* "slave": Send write queries to masters and read queries to slaves.

**For example, with `scaleReads: "slave"`:**

```js
cache.set('foo', 'bar') // This query will be sent to one of the masters.
cache.get('foo', (err, res) => {
  // This query will be sent to one of the slaves.
})
```

**Note:** In the code snippet above, the result may not be equal to "bar" because of the lag of replication between the master and slaves.

## Cache Fallback

In the case of a Redis connection failure, DADI Cache will attempt to reconnect four times before switching to file caching.
After a configurable period (default 5 minutes), an attempt will be made to reconnect to Redis and if successful DADI Cache will resume
using Redis as the cache store.

## Roadmap

* Add LATENCY HISTORY to assist in determining Redis performance
* Add FLUSH
* Add authentication options for Redis


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [2.0.0] - 2018-05-13

Add metadata methods to allow storing additional information with cached data

## [1.5.0] - 2018-04-13

Remove empty directories on cache flush

## [1.4.0] - 2017-03-05

Rewrite for cluster buffer and disabled reconnection

## [1.3.0] - 2016-12-29
### Changed
* Configuration properties `cleanseEnabled` and `cleanseInterval` have been renamed to `autoFlush` and `autoFlushInterval` respectively.

## [1.1.1] - 2016-10-12
### Added
- further configuration options documentation for file caching

## [1.1.1] - 2016-10-11
### Added
- retry strategy for Redis Cluster connections
- documentation for Redis Cluster hosts

## [1.1.0] - 2016-10-11
### Added
- full usage documentation

### Changed
- fixed the reconnection message output

## [1.0.0] - 2016-10-07
### Added
- this file
- travis-ci build file

### Changed
- set version to 1.0.0 in anticipation of release
- updated ioredis-mock to latest version
- ioredis-mock.expire & ioredis-mock.getrange now used instead of our own stub

## [0.0.1-Beta] - 2016-08-03
### Added
- project created
- ioredis used in place of node-redis



test/unit/file.js
var exec = require('child_process').exec
var fs = require('fs')
var fsExtra = require('fs-extra')
var path = require('path')
var should = require('should')
var sinon = require('sinon')
var Stream = require('stream')
var Cache = require(__dirname + '/../../lib/index')
var FileCache = require(__dirname + '/../../lib/file')

describe('FileCache', function () {
  it('should use empty string extension if none is specfied', function () {
    var cache = new Cache({ directory: { enabled: true, path: './cache' }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })
    var handler = cache.cacheHandler
    return handler.extension.should.eql('')
  })

  it('should use extension if one is specfied', function () {
    var cache = new Cache({ directory: { enabled: true, path: './cache', extension: 'json' }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })
    var handler = cache.cacheHandler
    return handler.extension.should.eql('.json')
  })

  describe('flush', function () {
    this.timeout(5000)

    beforeEach(function (done) {
      var setup = function (dir) {
        return new Promise((resolve, reject) => {
          exec('mkdir ' + dir, (err, y, z) => {
            // console.log(err)
            return resolve()
          })
        })
      }

      var cache = new Cache({ directory: { enabled: true, path: './cache' }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })
      setTimeout(() => {
        setup(path.resolve(cache.cacheHandler.directory)).then(done)
      }, 2000)
    })

    it('should use default autoFlush interval if none is specfied', function () {
      var cache = new Cache({ directory: { enabled: true, path: './cache' }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })

      cache.cacheHandler.autoFlushInterval.should.not.be.NaN()
      cache.cacheHandler.autoFlushInterval.should.be.above(0)
    })

    it('should use autoFlush interval if one is specfied', function () {
      var specificautoFlushInterval = 600
      var cache = new Cache({ directory: { enabled: true, path: './cache', autoFlushInterval: specificautoFlushInterval }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })

      cache.cacheHandler.autoFlushInterval.should.not.be.NaN()
      cache.cacheHandler.autoFlushInterval.should.equal(specificautoFlushInterval)
    })

    it('should remove an expired file when cache cleansing is enabled', function (done) {
      var cache = new Cache({ ttl: 1, directory: { enabled: true, path: './cache', autoFlush: true, autoFlushInterval: 1 }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })
      var filePath = path.resolve(path.join(cache.cacheHandler.directory, 'test_file'))
      fs.writeFileSync(filePath, 'testfile')

      setTimeout(() => {
        fs.stat(filePath, (err, stats) => {
          should.exist(err)
          should.not.exist(stats)
          done()
        })
      }, 2000)
    })

    it('should remove empty directories when cache cleansing is finished', function (done) {
      var cache = new Cache({ ttl: 1, directory: { enabled: true, path: './cache', autoFlush: true, autoFlushInterval: 1 }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })

      var dir = path.resolve(path.join(cache.cacheHandler.directory, 'test'))
      exec('mkdir -p ' + dir, (err, y, z) => {
        var filePath = dir + '/test_file'
        fs.writeFileSync(filePath, 'testfile')

        setTimeout(() => {
          fs.stat(dir, (err, stats) => {
            should.exist(err)
            should.not.exist(stats)
            done()
          })
        }, 2000)
      })
    })

    it('should not remove an unexpired file when cache cleansing is enabled', function (done) {
      var cache = new Cache({ ttl: 60, directory: { enabled: true, path: './cache', autoFlush: true, autoFlushInterval: 1 }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })
      var filePath = path.resolve(path.join(cache.cacheHandler.directory, 'test_file'))
      fs.writeFileSync(filePath, 'testfile')

      setTimeout(() => {
        fs.stat(filePath, (err, stats) => {
          should.not.exist(err)
          should.exist(stats)
          cache.cacheHandler.disableAutoFlush()
          done()
        })
      }, 2000)
    })
  })

  describe('set', function () {
    var cache = new Cache({ directory: { enabled: true, path: './cache', extension: 'json' }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })

    beforeEach(function (done) {
      var setup = function (dir) {
        return new Promise((resolve, reject) => {
          exec('mkdir ' + dir)
          return resolve()
        })
      }

      setup(path.resolve(cache.cacheHandler.directory)).then(done)
    })

    afterEach(function () {
      var files = fs.readdirSync(cache.cacheHandler.directory)
      files.forEach((file) => {
        try {
          fs.unlinkSync(cache.cacheHandler.directory + '/' + file)
        } catch (err) {}
      })
    })

    it('should generate a cache filename from the directory and key', sinon.test(function () {
      var spy = this.spy(fs, 'createWriteStream')
      cache.set('key1', 'data')

      return spy.firstCall.args[0].should.eql(path.resolve(cache.cacheHandler.directory + '/key1.json'))
    }))

    it('should create a cache file when a String is passed', function (done) {
      cache.set('key1', 'data')

      // check a file exists
      fs.stat(cache.cacheHandler.directory + '/key1.json', (err, stats) => {
        (!err).should.eql(true)
        done()
      })
    })

    it('should create a cache file when a Buffer is passed', function (done) {
      var buffer = new Buffer('data')
      cache.set('key1', buffer)

      // check a file exists
      fs.stat(cache.cacheHandler.directory + '/key1.json', (err, stats) => {
        (!err).should.eql(true)
        done()
      })
    })

    it('should create a cache file when a Stream is passed', function (done) {
      var stream = new Stream.Readable()
      stream.push('data')
      stream.push(null)
      cache.set('key1', stream)

      // check a file exists
      fs.stat(cache.cacheHandler.directory + '/key1.json', (err, stats) => {
        (!err).should.eql(true)
        done()
      })
    })
  })

  describe('metadata', function () {
    this.timeout(5000)

    let cache
    let metadata = {
      foo: 'bar',
      baz: 123
    }

    beforeEach(done => {
      cache = new Cache({
        directory: {
          enabled: true,
          path: './cache',
          extension: 'json'
        },
        redis: {
          enabled: false,
          host: '127.0.0.1',
          port: 6379
        }
      })

      setTimeout(done, 2000)
    })

    afterEach(() => {
      fsExtra.removeSync(
        path.join(__dirname, './../../db.json')
      )
    })

    it('should save metadata alongside a String payload', () => {
      return cache.set('key1', 'data', { metadata }).then(() => {
        return cache.get('key1')
      }).then(result => {
        should.exist(result)

        return cache.getMetadata('key1')
      }).then(result => {
        result.should.eql(metadata)
      })
    })

    it('should save metadata alongside a Buffer payload', () => {
      let buffer = new Buffer('data')

      return cache.set('key1', buffer, { metadata }).then(() => {
        return cache.get('key1')
      }).then(result => {
        should.exist(result)

        return cache.getMetadata('key1')
      }).then(result => {
        result.should.eql(metadata)
      })
    })

    it('should save metadata alongside a Stream payload', () => {
      let stream = new Stream.Readable()
      stream.push('data')
      stream.push(null)

      return cache.set('key1', stream, { metadata }).then(() => {
        return cache.get('key1')
      }).then(result => {
        return cache.getMetadata('key1')
      }).then(result => {
        result.should.eql(metadata)
      })
    })

    it('should persist metadata database to disk and reload it on instantiation', () => {
      return cache.set('key1', 'data', { metadata }).then(() => {
        return cache.get('key1')
      }).then(result => {
        return cache.getMetadata('key1')
      }).then(result => {
        result.should.eql(metadata)

        return new Promise((resolve, reject) => {
          setTimeout(resolve, 1500)
        }).then(() => {
          let newCache = new Cache({
            directory: {
              enabled: true,
              path: './cache',
              extension: 'json'
            },
            redis: {
              enabled: false,
              host: '127.0.0.1',
              port: 6379
            }
          })

          return cache.getMetadata('key1')
        }).then(result => {
          result.should.eql(metadata)
        })
      })
    })

    it('should delete any metadata associated with a key when it\'s flushed', done => {
      cache.set('key2', 'data', { metadata }).then(() => {
        return cache.get('key2')
      }).then(stream => {
        should.exist(stream)

        return cache.getMetadata('key2')
      }).then(result => {
        result.should.eql(metadata)

        return cache.flush('key*')
      }).then(() => {
        return cache.get('key2').catch(err => {
          err.message.should.eql(
            'The specified key does not exist'
          )

          return cache.getMetadata('key2').then(result => {
            should.equal(result, null)

            done()
          })
        })
      })
    })

    it('should not delete metadata associated with a key when another key is flushed', () => {
      cache.set('key3', 'data', { metadata }).then(() => {
        return cache.get('key3')
      }).then(stream => {
        should.exist(stream)

        return cache.getMetadata('key3')
      }).then(result => {
        result.should.eql(metadata)

        return cache.flush('key2')
      }).then(() => {
        return cache.get('key3')
      }).then(result => {
        result.should.eql(metadata)
      })
    })

    it('should return null when trying to get metadata for a key that doesn\'t have any', () => {
      return cache.getMetadata('key-that-does-not-exist').then(result => {
        should.equal(result, null)
      })
    })
  })

  describe('get', function () {
    var cache = new Cache({ directory: { enabled: true, path: './cache', extension: 'json' }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })

    after(function () {
      // remove cache folder contents completely, and recreate
      var cleanup = function (dir) {
        exec('rm -r ' + dir, function (err, stdout, stderr) {
          exec('mkdir ' + dir)
        })
      }

      cleanup(path.resolve(cache.cacheHandler.directory))
    })

    it('should generate a cache filename from the directory and key', sinon.test(function () {
      var spy = this.spy(fs, 'stat')

      cache.set('key1', 'data')
      cache.get('key1')

      return spy.firstCall.args[0].should.eql(path.resolve(cache.cacheHandler.directory + '/key1.json'))
    }))

    it('should generate a cache path with subdirectories when directoryChunkSize is set', sinon.test(function () {
      var cacheWithChunks = new Cache({ directory: { enabled: true, path: './cache', extension: 'json', directoryChunkSize: 4 }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })
      var spy = this.spy(fs, 'stat')

      var key = '1073ab6cda4b991cd29f9e83a307f34004ae9327'
      var expectedPath = path.resolve(cacheWithChunks.cacheHandler.directory + '/1073/ab6c/da4b/991c/d29f/9e83/a307/f340/04ae/9327' + '/1073ab6cda4b991cd29f9e83a307f34004ae9327.json')

      cacheWithChunks.set(key, 'data')

      cacheWithChunks.get(key)
      return spy.firstCall.args[0].should.eql(expectedPath)
    }))

    it('should generate a cache path with uneven-length subdirectories when directoryChunkSize is set', sinon.test(function () {
      var cacheWithChunks = new Cache({ directory: { enabled: true, path: './cache', extension: 'json', directoryChunkSize: 7 }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })
      var spy = this.spy(fs, 'stat')

      var key = '1073ab6cda4b991cd29f9e83a307f34004ae9327'
      var expectedPath = path.resolve(cacheWithChunks.cacheHandler.directory + '/1073ab6/cda4b99/1cd29f9/e83a307/f34004a/e9327' + '/1073ab6cda4b991cd29f9e83a307f34004ae9327.json')

      cacheWithChunks.set(key, 'data')

      cacheWithChunks.get(key)
      return spy.firstCall.args[0].should.eql(expectedPath)
    }))

    it('should reject if the key cannot be found', function (done) {
      cache.set('key1', 'data')
      cache.get('key2').then((stream) => {

      }).catch((err) => {
        err.message.should.eql('The specified key does not exist')
        done()
      })
    })

    it('should reject if the key has expired')

    it('should return a stream', function (done) {
      cache.set('key1', 'data')
      cache.get('key1').then((stream) => {
        stream.should.exist
        done()
      })
    })
  })
})


test/unit/cache.js
var should = require('should')
var Cache = require(__dirname + '/../../lib/index')

describe('Cache', function () {
  it('should allow instantiating a Cache instance', function () {
    var cache = new Cache()
    return cache.should.not.be.null
  })

  // it('should throw if invalid options are passed to the constructor', function () {
  //   return should.throws(function() { var cache = new Cache({ enabled: true  }) })
  // })

  it('should not be enabled if type == directory && enabled == false', function () {
    var cache = new Cache({ directory: { enabled: false, path: './cache' }, redis: { enabled: false } })
    return cache.enabled.should.eql(false)
  })

  // it('should throw if type == directory && path == null', function () {
  //   return should.throws(function() { var cache = new Cache({ type: 'directory', enabled: false }) })
  // })

  it('should be enabled if type == directory && enabled == true', function () {
    var cache = new Cache({ directory: { enabled: true, path: './cache' }, redis: { enabled: false } })
    return cache.enabled.should.eql(true)
  })

  it('should ignore falsy keys', function () {
    var cache = new Cache({ expireAt: null, directory: { enabled: true, path: './cache' }, redis: { enabled: false } })
    return cache.enabled.should.eql(true)
  })

  // it('should throw if type == redis && host == null', function () {
  //   return should.throws(function() { var cache = new Cache({ type: 'redis', enabled: true }) })
  // })

  // it('should throw if type == redis && port == null', function () {
  //   return should.throws(function() { var cache = new Cache({ type: 'redis', enabled: true, host: '127.0.0.1' }) })
  // })

  it('should be enabled if type == redis && enabled == true', function () {
    var cache = new Cache({ directory: { enabled: false }, redis: { enabled: true, host: '127.0.0.1', port: 6379 } })
    return cache.enabled.should.eql(true)
  })

  it('should create a FileCache if type == directory', function () {
    var cache = new Cache({ directory: { enabled: true, path: './cache' }, redis: { enabled: false, host: '127.0.0.1', port: 6379 } })
    return cache.cacheHandler.constructor.name.should.eql('FileCache')
  })

  it('should create a RedisCache if type == redis', function () {
    var cache = new Cache({ directory: { enabled: false, path: './cache' }, redis: { enabled: true, host: '127.0.0.1', port: 6379 } })
    return cache.cacheHandler.constructor.name.should.eql('RedisCache')
  })
})

lib/ioredis.js
const util = require('util')
const EventEmitter = require('events')
const redis = require('ioredis')
const redisRStream = require('redis-rstream')
const redisWStream = require('redis-wstream')
const Stream = require('stream')
const streamifier = require('streamifier')
const log = require('@dadi/logger')

function RedisCache (options) {
  this.options = options
  this.ttl = options.ttl || 3600
  this.redisClient = this.initialise(options)

  EventEmitter.call(this)
}

util.inherits(RedisCache, EventEmitter)

RedisCache.prototype.initialise = function initialise(options) {
  const self = this
  this.timesConnected = 0

  function retryStrategy(attempts) {
    const baseRetryTime = 1024
    const maxRetryTime = 4096
    const maxConnectedTimes = 3
    const currentRetryTime = baseRetryTime * attempts

    if (currentRetryTime > maxRetryTime) {
      self.onFailure()
      return new Error('Exceeded max retry time')
    }

    if (self.timesConnected > maxConnectedTimes) {
      self.onFailure()
      return new Error('Exceeded max times connected; Redis appears unstable')
    }

    return currentRetryTime
  }

  const redisClient = redis.createClient(options.port, options.host, {
    detect_buffers: true,
    retry_strategy
  })

  // error: doesn't get fired on dc errors
  redisClient.on('error', (err) => {
    log.error(err)
  })

  // end: should fire only on graceful dc
  redisClient.on('end', () => {
    this.emit('end')
  })

  // ready: when we are connected
  redisClient.on('ready', () => {
    this.timesConnected++
    this.emit('ready')
  })

  // reconnecting: every attempt
  redisClient.on('reconnecting', (attempt) => {
    this.emit('reconnecting', attempt)
  })

  return redisClient
}

RedisCache.prototype.get = function get(key) {
  return new Promise((resolve, reject) => {
    if (!this.redisClient.connected) {
      this.emit('fail')
      return reject(new Error('The specified key does not exist'))
    }

    this.redisClient.exists(key, (err, exists) => {
      if (exists > 0) {
        const stream = redisRStream(this.redisClient, key)
        return resolve(stream)
      } else {
        return reject(new Error('The specified key does not exist'))
      }
    })
  })
}

RedisCache.prototype.set = function set(key, data) {
  if (!this.redisClient.connected) {
    this.emit('fail', 'set', key, data)
  }

  return new Promise((resolve, reject) => {
    const redisWriteStream = redisWStream(this.redisClient, key)

    redisWriteStream.on('finish', () => {
      if (this.ttl) {
        this.redisClient.expire(key, this.ttl)
      }

      return resolve('')
    }).on('error', (err) => {
      return reject(err)
    })

    var stream

    // create a stream from the data if it is a String or Buffer
    if (data instanceof Buffer || typeof data === 'string') {
      stream = streamifier.createReadStream(data)
    } else if (data instanceof Stream) {
      stream = data
    }

    stream.pipe(redisWriteStream)
  })
}

RedisCache.prototype.onFailure = function onFailure() {
  const redisRetryTime = 1000 * 60 * 1

  this.emit('fail')

  // create an event to try redis again
  setTimeout(() => {
    log.warn('REDIS attempting reconnection')
    this.redisClient = this.initialise(this.options)
  }, redisRetryTime)
}

module.exports = function (options) {
  return new RedisCache(options)
}


scripts/coverage.js
#! /usr/bin/env node

var fs = require('fs')
var path = require('path')

var coberturaBadger = require('istanbul-cobertura-badger')

var opts = {
  badgeFileName: 'coverage',
  destinationDir: __dirname,
  istanbulReportFile: path.resolve(__dirname + '/../coverage', 'cobertura-coverage.xml'),
  thresholds: {
    excellent: 90, // overall percent >= excellent, green badge
    good: 60 // overall percent < excellent and >= good, yellow badge
  // overall percent < good, red badge
  }
}

// console.log(opts)

// Load the badge for the report$
coberturaBadger(opts, function parsingResults (err, badgeStatus) {
  if (err) {
    console.log('An error occurred: ' + err.message)
  }

  //console.log(badgeStatus)

  var readme = path.resolve(__dirname + '/../README.md')
  var badgeUrl = badgeStatus.url; // e.g. http://img.shields.io/badge/coverage-60%-yellow.svg

  // open the README.md and add this url
  fs.readFile(readme, {encoding: 'utf-8'}, function (err, body) {
    body = body.replace(/(!\[coverage\]\()(.+?)\?*(\))/g, function (whole, a, b, c) {
      return a + badgeUrl + c
    })

    fs.writeFile(readme, body, {encoding: 'utf-8'}, function (err) {
      if (err) console.log(err.toString())

      console.log('Coverage badge successfully added to ' + readme)
    })
  })
})


lib/index.js
const contract = require('dbc')
const EventEmitter = require('events')
const merge = require('deepmerge')
const path = require('path')
const FileCache = require('./file')
const RedisCache = require('./redis')
const util = require('util')

/**
 * Creates a new Cache instance
 * @constructor
 * @param {object} [options={ directory: { enabled: true, path: './cache' }, redis: { enabled: false } }] - the options used to instantiate Cache Handlers
 * @example
 * var Cache = require('@dadi/cache')
 * var cache = new Cache({
 *   "ttl": 3600,
 *   "directory": {
 *     "enabled": false,
 *     "path": "./cache/",
 *     "extension": "json"
 *   },
 *   "redis": {
 *     "enabled": true,
 *     "host": "127.0.0.1",
 *     "port": 6379
 *   }
 * })
 */
function Cache (options) {
  contract.check(options, {
    directory: [
      { validator: 'required' },
      { validator: 'type', args: ['object'] }
    ],
    redis: [
      { validator: 'required' },
      { validator: 'type', args: ['object'] }
    ]
  })

  this.options = options
  this.cacheHandlers = {}

  this.enabled = options.directory.enabled || options.redis.enabled

  // get the first cache type from the options with { enabled: true }
  this.type = Object.keys(options).filter(cacheHandler => {
    return options[cacheHandler] && options[cacheHandler].enabled
  })[0]

  switch (this.type) {
    case 'directory':
      this.cacheHandler = this.createFileCache(options.directory, options.ttl)
      break
    case 'redis':
      this.cacheHandler = this.createRedisCache(options.redis, options.ttl)
      break
    default:
      this.cacheHandler = this.createFileCache(options.directory, options.ttl)
  }
}

util.inherits(Cache, EventEmitter)

/**
 * Get an item from the cache
 * @param {String} key - the key used to reference the item in the cache
 * @returns {Promise.<Stream, Error>} A promise that returns a Stream if the key exists,
 *     or an Error if it does not exist
 */
Cache.prototype.get = function (key, options) {
  options = options || {}
  return this.cacheHandler.get(key, options)
}

/**
 * Get any metadata stored for a key
 * @param {String} key - the key used to reference the item in the cache
 * @returns {Promise.<Object, Error>} A promise that returns an object if there is metadata
 *     for the given key, or an Error otherwise
 */
Cache.prototype.getMetadata = function (key) {
  return this.cacheHandler.getMetadata(key)
}

/**
 * Add an item to the cache
 * @param {String} key - the key used to reference the item in the cache
 * @param {Buffer|Stream|String} data - the data to store in the cache
 * @returns {Promise.<String, Error>} A promise that returns an empty String if successful, otherwise an Error
 */
Cache.prototype.set = function (key, data, options) {
  options = options || {}
  return this.cacheHandler.set(key, data, options)
}

/**
 * Flush the cache
 * @param {String} path
 * @returns {Promise.<String, Error>} A promise that returns an empty String if successful, otherwise an Error
 */
Cache.prototype.flush = function (path) {
  return this.cacheHandler.flush(path)
}

/**
 * Instantiates a FileCache and adds it to the set of CacheHandlers
 * @param {object} options - the options used to create a FileCache instance, specifically `path` which determines where cached data will be stored
 * @param {Number} ttl - the time in seconds after which a cached item should be considered stale
 * @returns {FileCache}
 */
Cache.prototype.createFileCache = function (options, ttl) {
  let handler = new FileCache(Object.assign({}, options, {
    ttl
  }))

  this.cacheHandlers.directory = handler

  this.emit('message', 'FileCache connected')
  this.emit('ready')

  return handler
}

/**
 * Instantiates a RedisCache and adds it to the set of CacheHandlers
 * @param {object} options - the options used to create a RedisCache instance, specifically `host` and `port` for connecting to a Redis server
 * @param {Number} ttl - the time in seconds after which Redis should expire a cached item
 * @returns {RedisCache}
 */
Cache.prototype.createRedisCache = function (options, ttl) {
  let handler = new RedisCache(Object.assign({}, options, {
    ttl
  }))

  handler.on('ready', () => { // when we are connected
    if (this.cacheHandler.constructor.name === 'FileCache') {
      this.cacheHandler = this.cacheHandlers.redis
    }

    this.emit('message', 'REDIS connected')
    this.emit('ready')
  })

  handler.on('end', () => { // should fire only on graceful dc
    this.emit('message', 'REDIS disconnecting')
    this.cacheHandlers.redis = null
  })

  handler.on('reconnecting', function (attempt) { // every attempt
    this.emit('message', 'REDIS reconnecting, attempt #' + attempt)
  })

  handler.on('error', function (err) {
    this.emit('message', err)
  })

  handler.on('fail', (cmd, key, data) => {
    this.emit('message', 'REDIS connection failed. Falling back to filesystem caching at ' + path.resolve(this.options.directory.path))

    if (!this.cacheHandlers.directory) {
      this.createFileCache(this.options.directory, this.options.ttl)
    }

    delete this.cacheHandlers.redis

    this.cacheHandler = this.cacheHandlers.directory

    if (cmd === 'set') {
      return this.set(key, data)
    }

    if (cmd === 'get') {
      return this.get(key)
    }
  })

  this.cacheHandlers.redis = handler

  return handler
}

/**
 *
 */
module.exports = function (options) {
  var defaults = { directory: { enabled: false, path: 'cache' }, redis: { enabled: false } }
  var cacheOptions = merge(defaults, options || {})

  return new Cache(cacheOptions)
}
